Issue 524

abouroumine opened the pull request:
Tests

Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/blobber/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master BEFORE merging this PR

Associated PRs (Link as appropriate):
- zbox: https://github.com/0chain/zboxcli/pull/192
- zwalelet:
- gosdk: https://github.com/0chain/gosdk/pull/416
- 0chain:
- system_test: https://github.com/0chain/system_test/pull/218

abouroumine committed:
Issue 524

Tests

abouroumine committed:
Issue 524

Tests

abouroumine committed:
Systems Tests

Issue/524:
1- Done.
2- Done.
3- Not Yet.
4- Done.
5- Done.

abouroumine committed:
Systems Tests

Issue/524:
1- Done.
2- Done.
3- Not Yet.
4- Done.
5- Done.

abouroumine committed:
Issue/524

Step 3 Included.

abouroumine committed:
Issue/524

Step 3 Included.

abouroumine committed:
Issue/524

Fixed Unit Tests Error

abouroumine committed:
Issue/524

Fixed Unit Tests Error

abouroumine committed:
issue/524

Updating the Save for Directory.

abouroumine committed:
issue/524

Updating the Save for Directory.

abouroumine committed:
issue/524

New Update to the Steps 2 & 3.

abouroumine committed:
issue/524

New Update to the Steps 2 & 3.

abouroumine committed:
Issue/524

New Tests for steps 2-3

abouroumine committed:
Issue/524

New Tests for steps 2-3

abouroumine committed:
issue/524

removed dates to let sql them.
Step (3).

abouroumine committed:
issue/524

removed dates to let sql them.
Step (3).

abouroumine committed:
issue/524

Removed Step 2-3. Seems like the tests fails if not all Data is saved or if the Call to get all Data is performed.

abouroumine committed:
issue/524

Removed Step 2-3. Seems like the tests fails if not all Data is saved or if the Call to get all Data is performed.

abouroumine committed:
Merge branch 'issue/524' of https://github.com/0chain/blobber into issue/524

abouroumine committed:
Issue/524

Step 2-3 Update.

abouroumine committed:
Issue/524

Steps 2-3 Unit Tests Fixed.

abouroumine committed:
Issue/524

Step 3 Possible Fix

abouroumine committed:
issue/524

Possible fix to step 2-3

abouroumine committed:
Issue/524 => Step 3

Update of Step 3 of Issue/524

abouroumine committed:
Issue/524 Step 2

Fixing Step 2 of the Issue/524.

abouroumine committed:
issue/524

Enhancement to DB Optimization.

abouroumine committed:
Updates

abouroumine committed:
issue/524

Enhancement to the Solution proposed for DB Optimisation.

abouroumine committed:
issue/524

Updates

abouroumine committed:
issue/524

Enhancement to DB Optimisation.

abouroumine committed:
issue/524

Enhancement to the DB Optimisation.

abouroumine committed:
Merge branch 'staging' into issue/524

abouroumine committed:
Update file_changer_update.go

abouroumine committed:
Update file_changer_update.go

abouroumine committed:
Merge branch 'issue/524' of https://github.com/0chain/blobber into issue/524

abouroumine committed:
issue/524

DB Optimisation Enhancement.

abouroumine committed:
issue/524

DB Optimisation Enhancement Fixed lint error.

abouroumine committed:
issue/524

DB Optimisation.

abouroumine committed:
issue/524 (#554)

DB Optimisation.

abouroumine committed:
Update file_command_update.go

abouroumine committed:
Update file_command_delete.go

abouroumine committed:
issue/524

DB Optimisation Enhancement.

abouroumine committed:
Issue/524

DB Optimisation Enhancement. Step 3 not included (maybe better to keep it just Save).

abouroumine committed:
issue/524

Final Enhancement

abouroumine committed:
Merge branch 'staging' into issue/524

abouroumine committed:
issue/524

Unit Tests Conflict

abouroumine committed:
Merge branch 'issue/524' of https://github.com/0chain/blobber into issue/524

abouroumine committed:
issue/524

Fixed Tests Conflicts.

moldis commented on code/go/0chain.net/blobbercore/allocation/attributesfilechange.go:
```diff
@@ -27,7 +27,8 @@ func (ac *AttributesChange) ProcessChange(ctx context.Context, _ *AllocationChan
 
 	// root reference
-	ref, err = reference.GetReferencePath(ctx, ac.AllocationID, ac.Path)
+	ref, err = reference.GetReferencePath2(ctx, ac.AllocationID, ac.Path)
```
Name of function seems wrong

abouroumine replied:
This is a new function in code/go/0chain.net/blobbercore/reference/referencepath.go in line 25.

lpoli replied:
This is fine by me

moldis commented on code/go/0chain.net/blobbercore/allocation/file_changer_update.go:
```diff
@@ -90,7 +92,7 @@ func (nf *UpdateFileChanger) ProcessChange(ctx context.Context, change *Allocati
 			"setting file attributes: %v", err)
 	}
-
+	fmt.Println("Calculate Hash Need To Update Data !!!")
```
This logs are not relevant here, please make it more detail or just remove

abouroumine replied:
indeed there are some left logs that I reused and were not removed yet.

moldis commented on code/go/0chain.net/blobbercore/allocation/renamefilechange_test.go:
```diff
@@ -159,9 +159,10 @@ func TestBlobberCore_RenameFile(t *testing.T) {
 				mocket.Catcher.Reset()
-
-				query := `SELECT * FROM "reference_objects" WHERE ("reference_objects"."allocation_id" = $1 AND "reference_objects"."path" = $2 OR (path LIKE $3 AND allocation_id = $4)) AND "reference_objects"."deleted_at" IS NULL ORDER BY level, lookup_hash%!!(string=allocation id)!(string=/old_dir/%!)(MISSING)!(string=/old_dir)(EXTRA string=allocation id)`
+				//query := `SELECT "id","allocation_id","type","name","path","parent_path","size","hash","path_hash","content_hash","merkle_root","actual_file_size","actual_file_hash","attributes","chunk_size","lookup_hash" FROM "reference_objects" WHERE ("reference_objects"."allocation_id" = $1 AND "reference_objects"."path" = $2 OR (path LIKE $3 AND allocation_id = $4)) AND "reference_objects"."deleted_at" ...
```
No comments in the code, especially in PR for review

lpoli commented on code/go/0chain.net/blobbercore/handler/file_command_add.go:
```diff
@@ -36,7 +36,8 @@ func (cmd *AddFileCommand) IsAuthorized(ctx context.Context, req *http.Request,
 	}
-	exisitingFileRef, _ := reference.GetReference(ctx, allocationObj.ID, fileChanger.Path)
+	// Update GetReference to GetReferenceID
+	exisitingFileRef, _ := reference.GetReferenceID(ctx, allocationObj.ID, fileChanger.Path)
```
I think its better to atleast log error if any. In any case of err other than `gorm.ErrRecordNotFound` we should log actual error and probably return `common.NewError("some_code", "database error")

lpoli replied:
@abouroumine 
I am not sure about using GetReferenceID. Have you checked it with `ProcessContent` function as well ?
In other `file_command*` code `cmd.existingFileRef` is used which infers that we might need to consider using `reference.GetReference`.

Above all, I think its best to have `single responsibility principle`. That's why `cmd.existingFileRef` should have been assigned before calling `IsAuthorize`.

Please make changes for `IsAuthorize` function to have single responsibility.

abouroumine replied:
I think unlike in file_command_update / file_command_delete, in file_command_add we are just checking if the File Ref exists. and we do not use it in ProcessContent. that is why GetReferenceID was used instead of GetReference.
Now that you have mentioned we can even reduce the columns selected for other file_command_* depending on the columns used for each command in each ProcessContent and other related functions.

lpoli replied:
@abouroumine Its fine that you've called `GetReferenceID`. My point is that it is called inside `IsAuthorized` function making it hard to debug later on. We can call for existingRef just before `IsAuthorized` function is called and check for `ref` availability.

lpoli commented on code/go/0chain.net/blobbercore/handler/file_command_delete.go:
```diff
@@ -28,6 +28,7 @@ func (cmd *FileCommandDelete) IsAuthorized(ctx context.Context, req *http.Reques
 		return common.NewError("invalid_parameters", "Invalid path")
 	}
+	// Update GetReference to GetReferenceID
 	cmd.exisitingFileRef, _ = reference.GetReference(ctx, allocationObj.ID, path)
```
https://github.com/0chain/blobber/pull/529/files#r816421915

lpoli commented on code/go/0chain.net/blobbercore/handler/file_command_update.go:
```diff
@@ -42,7 +42,7 @@ func (cmd *UpdateFileCommand) IsAuthorized(ctx context.Context, req *http.Reques
 	}
-
+	// Update GetReference to GetReferenceID
 	cmd.exisitingFileRef, _ = reference.GetReference(ctx, allocationObj.ID, cmd.fileChanger.Path)
```
https://github.com/0chain/blobber/pull/529/files#r816421915

lpoli commented on code/go/0chain.net/blobbercore/handler/handler.go:
```diff
@@ -64,7 +64,7 @@ func SetupHandlers(r *mux.Router) {
 	r.HandleFunc("/v1/file/referencepath/{allocation}", common.ToJSONResponse(WithReadOnlyConnection(ReferencePathHandler)))
 	r.HandleFunc("/v1/file/objecttree/{allocation}", common.ToJSONResponse(WithReadOnlyConnection(ObjectTreeHandler)))
-	r.HandleFunc("/v1/file/refs/{allocation}", common.ToJSONResponse(WithReadOnlyConnection(RefsHandler))).Methods(http.MethodGet, http.MethodOptions)
+	r.HandleFunc("/v1/file/refs/{allocation}", common.ToJSONResponse(WithReadOnlyConnection(RefsHandler))).Methods(http.MethodGet, http.MethodOptions) // no db optimization yet.
```
Please check for this comment you have left

lpoli commented on config/0chain_blobber.yaml:
```diff
@@ -41,7 +41,7 @@ write_lock_timeout: 1m
 
 # delegate wallet (must be set)
-delegate_wallet: '2f34516ed8c567089b7b5572b12950db34a62a07e16770da14b15b170d0d60a9'
+delegate_wallet: 'ffb5f45e78cefc18e5b1c3fcdc9b0c87450dedc38d48abc7589e8f3af41a56c2'
```
I think we don't need this change. You may have done this for manual testing.

lpoli commented on config/0chain_validator.yaml:
```diff
@@ -1,7 +1,7 @@
 
 # delegate wallet (must be set)
-delegate_wallet: '86b147de5be951a5ab0c6b1732596686185c5e11c3bdf76a112183f828f39ba1'
+delegate_wallet: 'ffb5f45e78cefc18e5b1c3fcdc9b0c87450dedc38d48abc7589e8f3af41a56c2'
```
https://github.com/0chain/blobber/pull/529/files#r816429110

lpoli commented on code/go/0chain.net/blobbercore/handler/object_operation_handler.go:
```diff
@@ -474,18 +472,16 @@ func (fsh *StorageHandler) CommitWrite(ctx context.Context, r *http.Request) (*b
 		return nil, err
 	}
-	rootRef, err := reference.GetReference(ctx, allocationID, "/")
+	rootRef, err := reference.GetReferenceHash(ctx, allocationID, "/")
```
@abouroumine `connectionObj.ApplyChanges` can also return same rootHash. Can we do that?

lpoli replied:
@abouroumine Did you consider this?

lpoli commented on code/go/0chain.net/blobbercore/handler/object_operation_handler.go:
```diff
@@ -693,7 +690,8 @@ func (fsh *StorageHandler) UpdateObjectAttributes(ctx context.Context, r *http.R
 	var ref *reference.Ref
-	ref, err = reference.GetReferenceFromLookupHash(ctx, alloc.ID, pathHash)
+	// Update GetReferenceFromLookupHash to GetOnlyReferencePathFromLookupHash
+	ref, err = reference.GetOnlyReferencePathFromLookupHash(ctx, alloc.ID, pathHash)
```
Would `GetReferencePathFromLookupHash` be better than `GetOnlyReferencePathFromLookupHash`?

lpoli commented on code/go/0chain.net/blobbercore/handler/object_operation_handler.go:
```diff
@@ -693,7 +690,8 @@ func (fsh *StorageHandler) UpdateObjectAttributes(ctx context.Context, r *http.R
 	var ref *reference.Ref
-	ref, err = reference.GetReferenceFromLookupHash(ctx, alloc.ID, pathHash)
+	// Update GetReferenceFromLookupHash to GetOnlyReferencePathFromLookupHash
+	ref, err = reference.GetOnlyReferencePathFromLookupHash(ctx, alloc.ID, pathHash)
```
Would function name `GetReferencePathFromLookupHash` be better than `GetOnlyReferencePathFromLookupHash`?

abouroumine replied:
Both Functions are used in different places. The GetOnlyReferencePathFromLookupHash function selects only "id","path" where GetReferencePathFromLookupHash selects "id","name","path","hash","size","merkle_root". What I tried here is reduce the number of columns selected in Database query througth these two functions

lpoli replied:
@abouroumine Okay then would it be better if we do same thing as you did with `GetReferencePath2`?
In this case, `GetReferencePathFromLookupHash2` and provide doc string in the function as well

lpoli replied:
@abouroumine This changes may be missed. Please check this

lpoli commented on code/go/0chain.net/blobbercore/handler/object_operation_handler.go:
```diff
@@ -822,7 +822,8 @@ func (fsh *StorageHandler) DeleteFile(ctx context.Context, r *http.Request, conn
 
-	fileRef, _ := reference.GetReference(ctx, connectionObj.AllocationID, path)
+	//fileRef, _ := reference.GetReference(ctx, connectionObj.AllocationID, path)
+	fileRef, _ := reference.GetReferenceDelete(ctx, connectionObj.AllocationID, path)
```
@abouroumine Please don't ignore errors. Atleast log it.

lpoli commented on code/go/0chain.net/blobbercore/handler/object_operation_handler.go:
```diff
@@ -822,7 +822,8 @@ func (fsh *StorageHandler) DeleteFile(ctx context.Context, r *http.Request, conn
-	fileRef, _ := reference.GetReference(ctx, connectionObj.AllocationID, path)
+	//fileRef, _ := reference.GetReference(ctx, connectionObj.AllocationID, path)
+	fileRef, _ := reference.GetReferenceDelete(ctx, connectionObj.AllocationID, path)
 	_ = ctx.Value(constants.ContextKeyClientKey).(string)
```
Remove this redundant line#827 as well

lpoli commented on code/go/0chain.net/blobbercore/handler/object_operation_handler.go:
```diff
@@ -875,7 +875,8 @@ func (fsh *StorageHandler) CreateDir(ctx context.Context, r *http.Request) (*blo
 
-	exisitingRef := fsh.checkIfFileAlreadyExists(ctx, allocationID, dirPath)
+	// Update checkIfFileAlreadyExists to checkIfFileRefIDAlreadyExists
+	exisitingRef := fsh.checkIfFileRefIDAlreadyExists(ctx, allocationID, dirPath)
```
We can directly call `reference.GetReferenceID(ctx, allocationID, path)`. No need to call `checkIfFileRefIDAlreadyExists`.

lpoli commented on code/go/0chain.net/blobbercore/handler/object_operation_handler.go:
```diff
@@ -910,9 +911,9 @@ func (fsh *StorageHandler) CreateDir(ctx context.Context, r *http.Request) (*blo
 	formData.ActualHash = "-"
 	formData.ActualSize = 1
-
+	// We Need to do a Proper Look here
```
@abouroumine What about this comment?

abouroumine replied:
The problems revolving around those comments already resolved. just forgot to remove comments.

lpoli commented on code/go/0chain.net/blobbercore/handler/object_operation_handler.go:
```diff
@@ -1035,14 +1037,27 @@ func getFileOperation(r *http.Request) string {
 	return mode
 }
 
-func getExistingFileRef(fsh *StorageHandler, ctx context.Context, r *http.Request, allocationObj *allocation.Allocation, fileOperation string) *reference.Ref {
```
Please Delete commented code.

lpoli commented on code/go/0chain.net/blobbercore/handler/object_operation_handler.go:
```diff
@@ -1035,14 +1037,27 @@ func getFileOperation(r *http.Request) string {
 	if fileOperation == constants.FileOperationInsert || fileOperation == constants.FileOperationUpdate {
 		var formData allocation.UpdateFileChanger
 		uploadMetaString := r.FormValue(getFormFieldName(fileOperation))
 		err := json.Unmarshal([]byte(uploadMetaString), &formData)
```
@abouroumine If err is not nil then it is better to return error rather than fileRef being nil.
I think its better to add `error` in return and return this error to the client

lpoli commented on code/go/0chain.net/blobbercore/reference/objectpath.go:
```diff
@@ -19,6 +19,7 @@ type ObjectPath struct {
 
 // TODO needs to be refactored, current implementation can probably be heavily simplified
 func GetObjectPath(ctx context.Context, allocationID string, blockNum int64) (*ObjectPath, error) {
+	// Update GetRefWithSortedChildren to GetRefWithSortedChildrenForObjectPath
```
@abouroumine Address this comment

lpoli commented on code/go/0chain.net/blobbercore/reference/objectpath.go:
```diff
@@ -66,6 +67,7 @@ func GetObjectPath(ctx context.Context, allocationID string, blockNum int64) (*O
 				curRef = child
 				break
 			}
+			// Update GetRefWithSortedChildren to GetRefWithSortedChildrenForObjectPath
```
This as well

lpoli commented on code/go/0chain.net/blobbercore/reference/ref.go:
```diff
@@ -280,7 +381,8 @@ func GetRefWithSortedChildren(ctx context.Context, allocationID, path string) (*
 	db = db.Where(Ref{ParentPath: path, AllocationID: allocationID}).Or(Ref{Type: DIRECTORY, Path: path, AllocationID: allocationID})
-	err := db.Order("level, lookup_hash").Find(&refs).Error
+	//err := db.Order("level, lookup_hash").Find(&refs).Error
+	err := db.Order("level, path").Find(&refs).Error
```
@abouroumine I think its more appropriate to do `db.Order("path, level")`.

lpoli commented on code/go/0chain.net/blobbercore/reference/ref.go:
```diff
@@ -324,55 +426,55 @@ func (fr *Ref) GetFileHashData() string {
 	if len(r.Children) == 0 && !r.childrenLoaded {
 		return r.Hash, nil
 	}
 	sort.SliceStable(r.Children, func(i, j int) bool {
```
@abouroumine I think if we do `db.Order("path", "level")` then this sorting is redundant. Do you think so? 
Also the comparison it makes with `LookupHash` is questionable. We cannot order path with lookuphashes

abouroumine replied:
I think you are right the sorting is indeed redundant. and can you please explain more the difference between db.Order("path","level") than db.Order("level","path").

lpoli replied:
@abouroumine Now I think `level` is redundant. But the point was that sql ordering is heavy operation. First it gets all the rows and orders by some field and in that ordered result it orders again.
Also I think the difference between them is huge. In db.Order("path", "level") it first orders by path and then by level. In db.Order("level", "path") it first orders by level and then by path. 
In this two case they will have different set of ordered result.

`path` is unique and is sortable. `level` is not unique. So for example, `/a/b/c`, `/d/e/f` and `/d/e` are three paths.

If `db.Order('path', 'level')` then it results in:
/a/b/c
/d/e
/d/e/f

If `db.Order('level', 'path')` then it results in:
/d/e
/a/b/c
/d/e/f

lpoli commented on code/go/0chain.net/blobbercore/reference/ref.go:
```diff
@@ -422,6 +524,80 @@ func DeleteReference(ctx context.Context, refID int64, pathHash string) error {
 	return db.Where("path_hash = ?", pathHash).Delete(&Ref{ID: refID}).Error
 }
 
+func (r *Ref) SaveFile(ctx context.Context) error {
```
@abouroumine Please rename this function to something like `SaveFileRef` or `SaveFileRow`. Also same for `SaveDir`

lpoli commented on code/go/0chain.net/blobbercore/reference/ref.go:
```diff
@@ -324,55 +426,55 @@ func (fr *Ref) GetFileHashData() string {
 	var refNumBlocks int64
 	var size int64
 	for index, childRef := range r.Children {
+		_, err := childRef.CalculateHash(ctx, saveToDB)
```
@abouroumine I think `childRef.CalculateHash` is required only for affected refs. For unaffected refs we can simply get hash value from `Hash` field.

We should avoid hash calculation where we can because it is computationally resource consuming.
It would be great to distinguish affected and unaffected rows. For example if path is `/a/b/c/d/e.txt` then hash calculation is basically required for refs with paths `/`, `/a`, `/a/b`, `/a/b/c`, `/a/b/c/d` and `/a/b/c/d/e.txt`. Correct me If I am wrong here. @abouroumine @moldis

abouroumine replied:
Should we then introduce a new parameter that identifies the Affected and Unaffected Refs ? Especially if we are dealing with a file update / copy / move.

lpoli replied:
To identify affected rows new parameter is not required. Maybe it is required have a flag parameter to whether save or not save into database

lpoli commented on code/go/0chain.net/blobbercore/reference/referencepath.go:
```diff
@@ -22,6 +22,71 @@ func GetReferencePath(ctx context.Context, allocationID, path string) (*Ref, err
+			pathsAdded[path] = true
+		}
+		depth := len(GetSubDirsFromPath(path)) + 1
+		curPath := filepath.Dir(path)
```
This `curPath` prevails from previous code. I kind of don't like this logic. We could simply split path with `/` and get parent path by joining elements.
Something like:
```
splittedPaths := strings.Split("path", "/")
for i:=0;i<len(splittedPaths);i++{
    newPath := strings.Join(splittedPaths[0:i], "/")
}
```

abouroumine replied:
this update may trigger errors and wrong paths if the paths include "." as parent_path. and give wrong paths for the DB request. for now I may keep it until I figure out why the paths are not similar.

lpoli commented on code/go/0chain.net/blobbercore/reference/referencepath.go:
```diff
@@ -22,6 +22,71 @@ func GetReferencePath(ctx context.Context, allocationID, path string) (*Ref, err
+	// root reference_objects with parent_path=""
+	db = db.Or("parent_path = ? AND allocation_id = ?", "", allocationID)
+	//err := db.Order("level, lookup_hash").Find(&refs).Error
+	err := db.Order("level, path").Find(&refs).Error
```
Note Order("path", "level")

lpoli commented on code/go/0chain.net/blobbercore/reference/referencepath.go:
```diff
@@ -22,6 +22,71 @@ func GetReferencePath(ctx context.Context, allocationID, path string) (*Ref, err
+			refMap[refs[i].Path] = &refs[i]
+		}
+	}
+	if _, err := refs[0].CalculateHash(ctx, false); err != nil {
```
We may not need to calculate hash always when calling `GetReferenceForHashCalculationFromPaths`. Let's have a flag for whether to calculate hash or not.

lpoli commented on code/go/0chain.net/blobbercore/reference/referencepath.go:
```diff
@@ -105,12 +171,13 @@ func GetObjectTree(ctx context.Context, allocationID, path string) (*Ref, error)
 
-	err := db.Order("level, lookup_hash").Find(&refs).Error
+	//err := db.Order("level, lookup_hash").Find(&refs).Error
+	err := db.Order("level, path").Find(&refs).Error
```
Note Order("path", "level")

lpoli commented on code/go/0chain.net/blobbercore/handler/storage_handler.go:
```diff
@@ -124,8 +125,16 @@ func (fsh *StorageHandler) GetAllocationUpdateTicket(ctx context.Context, r *htt
+//	return fileReference
+//}
+
+func (fsh *StorageHandler) checkIfFileRefIDAlreadyExists(ctx context.Context, allocationID, path string) *reference.Ref {
```
Please put this function's body to previous function name

lpoli reviewed (changes requested):
@abouroumine 
Honestly I really appreciate your effort. I have commented in multiple places but as a sum up those changes are not much to work out.
Everyone has own method to solve problem. Though I would like to roughly explain my approach to this problem.

1. I would first create a postgres docker container and insert randomly generated millions rows with only necessary fields
2. Then calculate hashes and update all rows
3. After above step, make change in some row and re-calculate hash for the affected rows.
4. Calculate time taken so we can have performance difference.
5. Verify that calculation made by new function is correct.
6. Apply changes to blobber code base

abouroumine committed:
issue/524

first modifications related to PR comments.

abouroumine committed:
issue/524

new update regarding PR comments.

abouroumine committed:
Merge branch 'staging' into issue/524

abouroumine committed:
issue/524

Tests correction

abouroumine committed:
issue/524

Lints Tests errors fix.

abouroumine committed:
issue/524

Systems Tests fix

abouroumine committed:
issue/524

Hash Computation Optimisation.

abouroumine committed:
issue/524

lint tests fix. and hash computation system tests.

abouroumine committed:
issue/524

hash computation optimisation.

abouroumine committed:
issue/524

.

abouroumine committed:
issue/524

root hash computation.

abouroumine committed:
issue/524

Root Hash Calculation Optimisation.

abouroumine committed:
issue/524

gosdk version

abouroumine committed:
issue/524

gosdk update

abouroumine committed:
issue/524

Cleaned the PR for Review.

abouroumine committed:
issue/524

Cleaning the code for PR Review.

lpoli commented on code/go/0chain.net/blobbercore/reference/referencepath.go:
```diff
@@ -105,12 +167,11 @@ func GetObjectTree(ctx context.Context, allocationID, path string) (*Ref, error)
 	}
-
-	err := db.Order("level, lookup_hash").Find(&refs).Error
+	err := db.Order("path, level").Find(&refs).Error
```
Since path are unique fields, lets remove `level` here.

lpoli replied:
Since path are unique fields, lets remove `level` here.

lpoli commented on code/go/0chain.net/blobbercore/reference/referencepath.go:
```diff
@@ -88,14 +151,13 @@ func PathExists(ctx context.Context, allocationID, path string) (bool, error) {
 	var refs []Ref
 	db := datastore.GetStore().GetTransaction(ctx)
-
+	db = db.Select("id")
```
About this `PathExists` function, there is `IsRefExists` function that I recently merged into staging. Lets replace this function call by `IsRefExists`. 
`IsRefExists` is more optimized than `PathExists`

lpoli commented on code/go/0chain.net/blobbercore/reference/referencepath.go:
```diff
@@ -46,11 +109,11 @@ func GetReferencePathFromPaths(ctx context.Context, allocationID string, paths [
 	// root reference_objects with parent_path=""
 	db = db.Or("parent_path = ? AND allocation_id = ?", "", allocationID)
-	err := db.Order("level, lookup_hash").Find(&refs).Error
+	err := db.Order("path, level").Find(&refs).Error
```
Lets remove `level` as path is unique

lpoli commented on code/go/0chain.net/blobbercore/reference/referencepath.go:
```diff
@@ -18,8 +18,71 @@ type ReferencePath struct {
+func GetReferenceForHashCalculationFromPaths(ctx context.Context, allocationID string, paths []string) (*Ref, error) {
+	var refs []Ref
+	db := datastore.GetStore().GetTransaction(ctx)
+	db = db.Select("id", "allocation_id", "type", "name", "path",
```
Lets use gorm's smart select field. We can create a file for it where we can put smart select fields.

lpoli commented on code/go/0chain.net/blobbercore/reference/ref.go:
```diff
@@ -225,6 +224,109 @@ func GetReference(ctx context.Context, allocationID, path string) (*Ref, error)
+	return nil, err
+}
+
+func GetReferenceForCopyFromLookupHash(ctx context.Context, allocationID, path_hash string) (*Ref, error) {
```
This function is identical with `GetReferencePathFromLookupHash`. Can remove this function. We can add it later on if there are different parameters

lpoli commented on code/go/0chain.net/blobbercore/reference/ref.go:
```diff
@@ -225,6 +224,109 @@ func GetReference(ctx context.Context, allocationID, path string) (*Ref, error)
+	return nil, err
+}
+
+func GetOnlyReferencePathFromLookupHash(ctx context.Context, allocationID, path_hash string) (*Ref, error) {
```
`path_hash` should be `pathHash` as per go convention

peterlimg replied:
GetReferencePathByLookupHash() would be better, also path_hash to pathHash

lpoli commented on code/go/0chain.net/blobbercore/reference/ref.go:
```diff
@@ -225,6 +224,109 @@ func GetReference(ctx context.Context, allocationID, path string) (*Ref, error)
+}
+
+// GetReferenceHash get FileRef ID with allcationID and path from postgres
+func GetReferenceHash(ctx context.Context, allocationID, path string) (*Ref, error) {
```
If `hash` is only required then its better to return `hash` rather than `ref`. decoding to `ref` is costlier

lpoli commented on code/go/0chain.net/blobbercore/reference/referencepath.go:
```diff
@@ -18,8 +18,71 @@ type ReferencePath struct {
 
-func GetReferencePath(ctx context.Context, allocationID, path string) (*Ref, error) {
-	return GetReferencePathFromPaths(ctx, allocationID, []string{path})
+//func GetReferencePath(ctx context.Context, allocationID, path string) (*Ref, error) {
```
Remove commented code

lpoli commented on code/go/0chain.net/blobbercore/reference/referencepath.go:
```diff
@@ -18,8 +18,71 @@ type ReferencePath struct {
+			refMap[refs[i].Path] = &refs[i]
+		}
+	}
+	if _, err := refs[0].CalculateHash(ctx, false); err != nil {
```
if we are not storing it in db then do we even need to calculate hash? Its already there, right?

lpoli commented on code/go/0chain.net/blobbercore/allocation/newfilechange.go:
```diff
@@ -55,12 +55,12 @@ type NewFileChange struct {
 	tSubDirs := reference.GetSubDirsFromPath(path)
-
-	rootRef, err := reference.GetReferencePath(ctx, allocationID, nf.Path)
+	// Maybe Change this from GetReferencePath to GetReferencePath2
```
Remove this comment

lpoli reviewed (changes requested):
I only see few changes to be made. Otherwise it looks good to me

peterlimg commented on code/go/0chain.net/blobbercore/handler/file_command_add.go:
```diff
@@ -3,12 +3,12 @@ package handler
 import (
 	"context"
 	"encoding/json"
+	"github.com/0chain/blobber/code/go/0chain.net/blobbercore/reference"
```
run goimports to format

peterlimg commented on code/go/0chain.net/blobbercore/handler/file_command_add.go:
```diff
@@ -36,9 +38,11 @@ func (cmd *AddFileCommand) IsAuthorized(ctx context.Context, req *http.Request,
-	exisitingFileRef, _ := reference.GetReference(ctx, allocationObj.ID, fileChanger.Path)
-
-	if exisitingFileRef != nil {
+	cmd.existingFileRef, err = reference.GetReferenceID(ctx, allocationObj.ID, fileChanger.Path)
```
What's the purpose for saving this existingFileRef to the cmd, if the file is duplicate, we will return error, and it won't be used.

abouroumine replied:
to avoid redundant call to db for checking if file exists in both object_operation_handler => WriteFile function and file_command_add IsAuthorized.

peterlimg commented on code/go/0chain.net/blobbercore/handler/object_operation_handler.go:
```diff
@@ -590,7 +587,8 @@ func (fsh *StorageHandler) RenameObject(ctx context.Context, r *http.Request) (i
 	defer mutex.Unlock()
 
-	objectRef, err := reference.GetReferenceFromLookupHash(ctx, allocationID, pathHash)
+	// Changed From GetReferenceFromLookupHash To GetReferenceForHashCalculationFromPaths
```
Incorrect comment

peterlimg commented on code/go/0chain.net/blobbercore/handler/object_operation_handler.go:
```diff
@@ -947,8 +948,18 @@ func (fsh *StorageHandler) WriteFile(ctx context.Context, r *http.Request) (*blo
-	fileOperation := getFileOperation(r)
-	existingFileRef := getExistingFileRef(fsh, ctx, r, allocationObj, fileOperation)
+	cmd := createFileCommand(r)
+	err2 := cmd.IsAuthorized(ctx, r, allocationObj, clientID)
```
Check err2 right after it is returned to avoid unnecessary code checking, also, use `err` instead of `err2`

abouroumine replied:
the reason I'm checking the err2 later is due to multiple system, unit and integration tests failures. since we are calling the IsAuthorized much ahead now to reduce db requests.

peterlimg replied:
If we got error from `IsAuthorized()` and do not check and return early, the execution of the following `IsACollaborator()`, `verifySignatureFromRequest()` would be a waste of time/resources, and moving `IsAuthorized()` ahead would not make sense. 

We should update unit tests and system tests instead of posting the the error check.

peterlimg replied:
Please fix this

peterlimg commented on code/go/0chain.net/blobbercore/handler/storage_handler.go:
```diff
@@ -124,12 +124,8 @@ func (fsh *StorageHandler) GetAllocationUpdateTicket(ctx context.Context, r *htt
-		return nil
-	}
-	return fileReference
+func (fsh *StorageHandler) checkIfFileRefIDAlreadyExists(ctx context.Context, allocationID, path string) (*reference.Ref, error) {
```
No  need  to change function name

peterlimg commented on code/go/0chain.net/blobbercore/reference/ref.go:
```diff
@@ -225,6 +224,109 @@ func GetReference(ctx context.Context, allocationID, path string) (*Ref, error)
+	return nil, err
+}
+
+// GetReferenceDelete get FileRef ID with allcationID and path from postgres
```
Please comment appropriately

peterlimg commented on code/go/0chain.net/blobbercore/reference/ref.go:
```diff
@@ -225,6 +224,109 @@ func GetReference(ctx context.Context, allocationID, path string) (*Ref, error)
+}
+
+// GetReferenceDelete get FileRef ID with allcationID and path from postgres
+func GetReferenceDelete(ctx context.Context, allocationID, path string) (*Ref, error) {
```
What does the `Delete` mean in the function name, I don't see any delete relate action

peterlimg commented on code/go/0chain.net/blobbercore/reference/ref.go:
```diff
@@ -225,6 +224,109 @@ func GetReference(ctx context.Context, allocationID, path string) (*Ref, error)
+	return nil, err
+}
+
+func GetReferencePathFromLookupHash(ctx context.Context, allocationID, path_hash string) (*Ref, error) {
```
GetReferenceByLookupHash(), it does not only return path

peterlimg commented on code/go/0chain.net/blobbercore/reference/referencepath.go:
```diff
@@ -18,8 +18,71 @@ type ReferencePath struct {
+//	return GetReferencePathFromPaths(ctx, allocationID, []string{path})
+//}
+
+func GetReferencePath2(ctx context.Context, allocationID, path string) (*Ref, error) {
```
Rename to GetReferencePath if it is used.

peterlimg commented on code/go/0chain.net/blobbercore/stats/filestats.go:
```diff
@@ -2,7 +2,6 @@ package stats
 import (
 	"context"
-
 	"github.com/0chain/blobber/code/go/0chain.net/blobbercore/datastore"
```
run  goimports to format

abouroumine committed:
issue/524

PR Review updates.

abouroumine committed:
Merge branch 'staging' into issue/524

abouroumine committed:
issue/524

PR Review merged.

abouroumine committed:
issue/524

PR Review Fix.

abouroumine committed:
issue/524

Lint Test fix

abouroumine committed:
issue/524

PR Review fix

abouroumine committed:
issue/524

PR Review fix.

abouroumine committed:
issue/524

PR Review Finished.

abouroumine committed:
issue/524

PR Review Finished. goimports

abouroumine committed:
issue/524

lookup_hash removal.

abouroumine committed:
issue/524

abouroumine committed:
Merge branch 'staging' into issue/524

abouroumine committed:
issue/524

PR Review Fix

abouroumine committed:
issue/524

go.sum fix

abouroumine committed:
issue/524

updated zbox/gosdk related to this blobber.

service-0chain commented:
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | blobber/issue-524  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | issue/524  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

abouroumine committed:
issue/524

fixed some system tests related to Delete & LoadObjectTree.

service-0chain commented:
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | blobber/issue-524  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | issue/524  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

abouroumine committed:
issue/524

cleaned for review.

lpoli commented on code/go/0chain.net/blobbercore/datastore/postgres.go:
```diff
@@ -72,7 +72,7 @@ func (store *postgresStore) GetDB() *gorm.DB {
 
 func (store *postgresStore) AutoMigrate() error {
-
+	time.Sleep(20 * time.Second)
```
Please remove this

lpoli commented on code/go/0chain.net/blobbercore/handler/object_operation_handler.go:
```diff
@@ -783,12 +778,16 @@ func (fsh *StorageHandler) CopyObject(ctx context.Context, r *http.Request) (int
 	}
 	newPath := filepath.Join(destPath, objectRef.Name)
+	destRef, _ := reference.GetLimitedRefFieldsByPath(ctx, allocationID, newPath, []string{"id"})
+	if destRef != nil {
```
I think we should remove this because this is already checked in Line 807 below.

abouroumine replied:
it did cause some system tests errors due to the order of the errors handling.

lpoli commented on code/go/0chain.net/blobbercore/reference/object.go:
```diff
@@ -17,9 +17,10 @@ func LoadObjectTree(ctx context.Context, allocationID, path string) (*Ref, error
 
-	db = db.Order("level desc, lookup_hash")
+	//db = db.Order("level desc, lookup_hash")
+	db = db.Order("level desc, path")
```
Maybe `db.Order("path")` instead of `db.Order("level desc, path")` ?

abouroumine replied:
db = db.Order("path desc") or db = db.Order("path") result into allocation root error for some system tests. the order by level desc is very important here during the deletion process.

lpoli replied:
Yes its because of below logic L#29-L#40.
It adds child in backward direction.

lpoli reviewed (changes requested):
In one place you've used `db.Order("level desc, path")`. Please check for this.

abouroumine committed:
Update postgres.go

Review Updates.

abouroumine committed:
Merge branch 'staging' into issue/524

abouroumine committed:
Merge branch 'staging' into issue/524

abouroumine committed:
Update postgres.go

abouroumine committed:
issue/524

new gosdk merge.

service-0chain commented:
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | blobber/issue-524  |
| zboxcli  | blobber/issue-524  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | issue/524  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

abouroumine committed:
Update object.go

abouroumine committed:
Update object.go

up for review.

service-0chain commented:
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | blobber/issue-524  |
| zboxcli  | blobber/issue-524  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | issue/524  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

peterlimg commented on code/go/0chain.net/blobbercore/allocation/renamefilechange.go:
```diff
@@ -25,7 +25,12 @@ func (rf *RenameFileChange) DeleteTempFile() error {
+
+	isFilePresent, err := reference.IsRefExist(ctx, rf.AllocationID, rf.NewName)
+	if err != nil {
+		Logger.Info("invalid_reference_path", zap.Any("error", err))
```
Log error in this way:
```go
Logger.Error("invalid_reference_path", zap.Error(err))
```

peterlimg commented on code/go/0chain.net/blobbercore/handler/object_operation_handler.go:
```diff
@@ -834,17 +833,19 @@ func (fsh *StorageHandler) CopyObject(ctx context.Context, r *http.Request) (int
-	fileRef, _ := reference.GetReference(ctx, connectionObj.AllocationID, path)
+	fileRef, err := reference.GetLimitedRefFieldsByPath(ctx, connectionObj.AllocationID, path, []string{"path", "name", "size", "hash", "merkle_root"})
+	if err != nil {
+		Logger.Info("invalid_file", zap.Any("error", err))
```
Log error with Logger.Error(.., zap.Error(err))

peterlimg commented on code/go/0chain.net/blobbercore/handler/object_operation_handler.go:
```diff
@@ -897,11 +897,15 @@ func (fsh *StorageHandler) CreateDir(ctx context.Context, r *http.Request) (*blo
 
+	exisitingRef, err := fsh.checkIfFileAlreadyExists(ctx, allocationID, dirPath)
+	if err != nil {
+		Logger.Info("Error file reference", zap.Any("error", err))
```
Logger.Error(.., zap.Error(err))

peterlimg commented on code/go/0chain.net/blobbercore/reference/ref.go:
```diff
@@ -225,36 +223,38 @@ func GetReference(ctx context.Context, allocationID, path string) (*Ref, error)
+	db = db.Select(selectedFields)
+	err := db.Where(&Ref{AllocationID: allocationID, LookupHash: lookupHash}).First(ref).Error
+	if err == nil {
+		return ref, nil
```
Please handle the error in the Go idiomatic way, i.e. check if `err != nil` to return error early

peterlimg commented on code/go/0chain.net/blobbercore/reference/ref.go:
```diff
@@ -225,36 +223,38 @@ func GetReference(ctx context.Context, allocationID, path string) (*Ref, error)
-		return nil, err
+	err := db.Where(&Ref{AllocationID: allocationID, LookupHash: pathHash}).First(ref).Error
+	if err == nil {
+		return ref, nil
```
Check and return error early

peterlimg commented on code/go/0chain.net/blobbercore/allocation/attributesfilechange.go:
```diff
@@ -38,13 +39,14 @@ func (ac *AttributesChange) ApplyChange(ctx context.Context, _ *AllocationChange
 		treelevel = 0
 	)
-
+	dirRef.HashToBeComputed = true
```
What's rule for setting this `HashToBeComputed`, i.e when should we set it to true, and when to false

abouroumine replied:
For the HashToBeComputed flag parameter is the result of this: https://github.com/0chain/blobber/pull/529#discussion_r816455077 
since whenever there's new/update/delete file we have to update the hash but we can use hashes that did not change to optimise the calculations.

abouroumine committed:
issue/524

new update

abouroumine committed:
Merge branch 'staging' into issue/524

abouroumine committed:
issue/524

updates tests.

abouroumine committed:
Update handler_test.go

service-0chain commented:
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | blobber/issue-524  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | blobber/issue-524  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | issue/524  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

service-0chain commented:
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | blobber/issue-524  |
| gosdk  | blobber/issue-524  |
| zboxcli  | blobber/issue-524  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | issue/524  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

service-0chain commented:
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | blobber/issue-524  |
| gosdk  | blobber/issue-524  |
| zboxcli  | blobber/issue-524  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | issue/524  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

abouroumine committed:
Merge branch 'staging' into issue/524

abouroumine committed:
issue/524

update go ( mod/sum ) + unit tests.

abouroumine committed:
Update handler_test.go

abouroumine committed:
issue/524

update depending on the https://github.com/0chain/blobber/pull/529/files#r831803817

peterlimg commented on code/go/0chain.net/blobbercore/handler/handler_test.go:
```diff
@@ -248,6 +235,15 @@ func isEndpointRequireSignature(name string) bool {
+		return true
+	default:
+		return false
+	}
```
Could be a simple `return if name == "Upload"` unless there are other possible names can be treated as Upload.

peterlimg commented on code/go/0chain.net/blobbercore/handler/handler_test.go:
```diff
@@ -35,6 +24,15 @@ import (
+	"github.com/gorilla/mux"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
```
What format tools are you using, we'd better run `goimports` so that all have the same code format

abouroumine replied:
Goland IDE => Code Style => Go => goimports.

abouroumine committed:
Update handler_test.go

abouroumine committed:
Update handler_test.go

abouroumine committed:
Merge branch 'staging' into issue/524

abouroumine committed:
Update object_operation_handler.go

abouroumine committed:
Update handler_test.go

abouroumine committed:
Update storage_handler.go

abouroumine committed:
Merge branch 'staging' into issue/524

abouroumine committed:
issue/524

Update goSDK mod/sum

peterlimg commented on code/go/0chain.net/blobbercore/handler/object_operation_handler.go:
```diff
@@ -809,17 +806,19 @@ func (fsh *StorageHandler) CopyObject(ctx context.Context, r *http.Request) (int
-	fileRef, _ := reference.GetReference(ctx, connectionObj.AllocationID, path)
+	fileRef, err := reference.GetLimitedRefFieldsByPath(ctx, connectionObj.AllocationID, path, []string{"path", "name", "size", "hash", "merkle_root"})
+	if err != nil {
+		Logger.Info("invalid_file", zap.Error(err))
```
Please log error with `Logger.Error()` as it will give stack info, so as other places.

abouroumine committed:
Update object_operation_handler.go

abouroumine committed:
Update object_operation_handler.go

peterlimg commented on code/go/0chain.net/blobbercore/handler/object_operation_handler.go:
```diff
@@ -952,8 +953,22 @@ func (fsh *StorageHandler) WriteFile(ctx context.Context, r *http.Request) (*blo
+		existingFileRef = rCmd.existingFileRef
+	case *FileCommandDelete:
+		existingFileRef = rCmd.existingFileRef
+	}
```
The `*UpdateFileCommand` and `*FileCommandDelete` cases can be handled in one case in the following way:

```go
switch rCmd := cmd.(type) {
...
case *UpdateFileCommand, 
        *FileCommandDelete:
    existingFileRef = rcmd.existingFileRef
...
}
```

Also, please return error if the type was unknown.

abouroumine replied:
I think the code if I update it I should do something like this:
```
        case *UpdateFileCommand:
		existingFileRef = rCmd.existingFileRef
	case *FileCommandDelete:
		existingFileRef = rCmd.existingFileRef
	default:
		existingFileRef = nil
```

- We can't get the `rCmd.existingFileRef` if we don't specify exactly the type createFileCommand of the `rCmd`.
If we do something like this: 
```
 case *UpdateFileCommand, 
        *FileCommandDelete:
    existingFileRef = rcmd.existingFileRef
```
the rcmd.existingFileRef is not recognized.
- The Error is handled before we reach this step in:
```
        err = cmd.IsValidated(ctx, r, allocationObj, clientID)
	if err != nil {
		return nil, err
	}
	var existingFileRef *reference.Ref
	switch rCmd := cmd.(type) {
	case *UpdateFileCommand:
		existingFileRef = rCmd.existingFileRef
	case *FileCommandDelete:
		existingFileRef = rCmd.existingFileRef
	default:
		existingFileRef = nil
	}
```
Here we just affect to the existingFileRef for Checking if possible Collaborator.

abouroumine committed:
Update object_operation_handler.go

peterlimg reviewed (approved):
lgtm

abouroumine committed:
Merge branch 'staging' into issue/524

abouroumine committed:
Merge branch 'staging' into issue/524

abouroumine committed:
issue/524

update goSDK

abouroumine committed:
Merge branch 'staging' into issue/524

abouroumine committed:
issue/524

update goSDK

service-0chain commented:
Manual system tests [cancelled] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | staging  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |
